File: 04_agent/claude-agent-sdk/src/tools/taskTools.ts
import { v4 as uuidv4 } from "uuid";

type Priority = "low" | "medium" | "high";
type Status = "todo" | "in_progress" | "done";

export type Task = {
  id: string;
  title: string;
  description: string;
  priority: Priority;
  status: Status;
  createdAt: string;
};

export const taskStore: Task[] = [];

type CreateInput = { title: string; description: string; priority: Priority };

export const createTask = (input: CreateInput): Task => {
  const task: Task = {
    id: uuidv4(),
    ...input,
    status: "todo",
    createdAt: new Date().toISOString(),
  };
  taskStore.push(task);
  return task;
};

export const listTasks = (): Task[] => [...taskStore];

export const updateTask = (
  id: string,
  updates: Partial<Omit<Task, "id" | "createdAt">>
): Task | null => {
  const idx = taskStore.findIndex((t) => t.id === id);
  if (idx === -1) return null;
  taskStore[idx] = { ...taskStore[idx], ...updates };
  return taskStore[idx];
};

export const deleteTask = (id: string): boolean => {
  const idx = taskStore.findIndex((t) => t.id === id);
  if (idx === -1) return false;
  taskStore.splice(idx, 1);
  return true;
};

File: 04_agent/strands-python/backend/tools/task_tools.py
from uuid import uuid4
from datetime import datetime, timezone
from strands import tool

task_store: list[dict] = []


@tool
def create_task(title: str, description: str, priority: str) -> dict:
    """新しいタスクを作成する。priorityはlow/medium/highのいずれか。"""
    task = {
        "id": str(uuid4()),
        "title": title,
        "description": description,
        "priority": priority,
        "status": "todo",
        "createdAt": datetime.now(timezone.utc).isoformat(),
    }
    task_store.append(task)
    return task


@tool
def list_tasks() -> list:
    """タスク一覧を取得する。"""
    return list(task_store)


@tool
def update_task(id: str, title: str = "", status: str = "", priority: str = "") -> dict:
    """タスクを更新する。更新したいフィールドのみ指定する。"""
    for task in task_store:
        if task["id"] == id:
            if title:
                task["title"] = title
            if status:
                task["status"] = status
            if priority:
                task["priority"] = priority
            return task
    return {}


@tool
def delete_task(id: str) -> bool:
    """タスクを削除する。"""
    for i, task in enumerate(task_store):
        if task["id"] == id:
            task_store.pop(i)
            return True
    return False
